File: templates/master/examples/js/Quiz.js
var _=require('lodash')
var aws=require('aws-sdk')
aws.config.region=process.env.AWS_REGION
var lambda=new aws.Lambda()
var kms=new aws.KMS()
var handlebars = require('handlebars')
var fs=require('fs')

var markdown=handlebars.compile(
    fs.readFileSync(`${__dirname}/templates/quiz-response.md`,'utf-8')
)
var text=handlebars.compile(
    fs.readFileSync(`${__dirname}/templates/quiz-response.hbs`,'utf-8')
)

exports.handler=async function(event,context,callback){
    console.log(JSON.stringify(event,null,2))
    try{
        if(event.res.session.quizBot){
            var decrypt=await kms.decrypt({
                CiphertextBlob:Buffer.from(event.res.session.quizBot,'base64'),
                EncryptionContext:{
                    userId:event.req._event.userId
                }
            }).promise()
            console.log(decrypt)
            var quizBot=JSON.parse(decrypt.Plaintext.toString('utf8'))
        }else{
            var quizBot={
                questionCount:0,
                correctAnswerCount:0,
                next:event.res.result.args[0],
                originalDocumentQid:_.get(event,"res.session.previous.qid","")
            }
        }
        var templateParams={
            first:quizBot.questionCount===0,
            message:_.get(event,"res.result.a")
        } 
        if(quizBot.questionCount>0){
            templateParams.correctAnswers=quizBot.correctAnswers
            if(isCorrect(event.req.question,quizBot.correctAnswers)){
                templateParams.correct=true
                quizBot.correctAnswerCount++
            }else{
                templateParams.incorrect=true
            }
        }
        
        if( !quizBot.next || ["QUIT","EXIT"].includes(standardize(quizBot.next)) ){
            templateParams.finished=true
            templateParams.totalCorrect=quizBot.correctAnswerCount
            templateParams.totalQuestions=quizBot.questionCount
            var score=quizBot.correctAnswerCount/quizBot.questionCount*100
            templateParams.score=Math.round(score)
            delete event.res.session.quizBot
            delete event.res.session.queryLambda
        }
        
        if(quizBot.next){
            result=await lambda.invoke({
                FunctionName:event.req._info.es.service.qid,
                InvocationType:"RequestResponse",
                Payload:JSON.stringify({qid:quizBot.next})
            }).promise()

            nextDocument=JSON.parse(result.Payload)
            console.log(JSON.stringify(nextDocument,null,2))
            if(!nextDocument) throw `Next Document not Found:${quizBot.next}` 

            templateParams.question=nextDocument.question 
            templateParams.answers=_.shuffle(
                nextDocument.incorrectAnswers.map(answer=>[answer,false])
                .concat(
                    nextDocument.correctAnswers.map(answer=>[answer,true])
                ))
                .map((val,index)=>{
                    val[2]=String.fromCharCode(65+index)
                    return val
                })
            
            quizBot.correctAnswers=templateParams.answers
                .filter(x=>x[1]).map(x=>x[2])

            event.res.session.queryLambda=process.env.AWS_LAMBDA_FUNCTION_NAME
            quizBot.questionCount++
            quizBot.next=_.get(nextDocument,"next[0]",false)
          
            var encrypt=await kms.encrypt({
                KeyId:process.env.QUIZ_KMS_KEY,
                Plaintext:JSON.stringify(quizBot),
                EncryptionContext:{ 
                    userId:event.req._event.userId
                }
            }).promise()
            console.log(encrypt)

            event.res.session.quizBot=encrypt.CiphertextBlob.toString('base64')
            if(_.get(nextDocument,"r.imageUrl")){
                event.res.card=nextDocument.r
                event.res.card.send=true
            }
        }

        event.res.message=text(templateParams)
            .replace(/\r?\n|\r/g, " ").replace(/ +(?= )/g,'');
        _.set(event,
            "res.session.appContext.altMessages.markdown",
            markdown(templateParams)
        )
        
        console.log(JSON.stringify(event,null,2))
        callback(null,event)
    }catch(e){
        console.log("Failed",e)
        delete event.res.session.quizBot
        delete event.res.session.queryLambda
        event.message="Sorry, Failed to process quiz"
        callback(null,event)
    }
}

function isCorrect(response,list){
    return list.includes(standardize(response))
}

function standardize(str){
    return str.toUpperCase().trim().replace(/[^\w\s]|_/g, "")
}

File: templates/master/examples/py/Quiz.py
import json
import boto3
import os
import botocore.response as br
import string
import random

def handler(event,context):
    print("Entering handler, received following event: ")
    print(json.dumps(event))
    
    botResponse = event["res"].get("message", "")
    quizBot=event["res"]["session"].get("quizBot",{}) 
    nextQuestionQid = ""
        
    if (!quizBot):
        nextQuestionQid = event["res"]["result"].get("args")[0]
        quizBot = {
            "questionCount":0,
            "correctAnswerCount":0,
            "next":nextQuestionQid,
            "originalDocumentQid":event["res"]["session"]["previous"].get("qid", "")
        }
        
    if (quizBot["questionCount"] > 0):
        quizBot["userAnswer"]=event["req"]["question"]
        correctAnswerIndices = quizBot["correctAnswerIndices"]
        userAnswerIndex = ord(standardizeString(quizBot["userAnswer"])[:1]) - ord('A')
        
        if (userAnswerIndex in correctAnswerIndices):
            quizBot["correctAnswerCount"] += 1
            botResponse += "Correct answer!"
        else:
            botResponse += "Sorry, that was incorrect. "
            if (len(correctAnswerIndices) == 1):
                botResponse += "The correct answer was {0}. ".format(chr(correctAnswerIndices[0] + ord('A')))
            else:
                botResponse += "All of the following answers were correct: "
                for x in correctAnswerIndices:
                    botResponse += chr(x + ord('A')) + ", "
                    
                # these lines help the bot's grammar
                botResponse = botResponse[:-3] + "and " + botResponse[-3:]
                botResponse = botResponse[:-2]
    
    currentDocument=getNextQuestionQid(quizBot)
    quizBot["current"]=currentDocument 
    if (currentDocument):
        currentDocumentContents = getDocumentContents(currentDocument, event)
        botResponse += "\n" + currentDocumentContents["question"] + " \n"
        
        answerPairs = assembleAnswerOptions(currentDocumentContents)
        quizBot["correctAnswerIndices"] = []
        for i, answerPair in enumerate(answerPairs):
            botResponse += answerPair[0] + " \n"
            if (answerPair[1]):
                quizBot["correctAnswerIndices"].append(i)
                
        if ("r" in currentDocumentContents):
            event["res"]["card"]={
                "send":True,
                "title":currentDocumentContents["r"]["title"],
                "imageUrl":currentDocumentContents["r"]["imageUrl"]
            }
        event["res"]["session"]["queryLambda"] = os.environ['AWS_LAMBDA_FUNCTION_NAME']
        quizBot["questionCount"] += 1
        quizBot["next"] = getNextQuestionQid(currentDocumentContents)
    if not quizShouldContinue(event):
        questionsAnsweredCount = quizBot["questionCount"]
        questionsCorrectCount = quizBot["correctAnswerCount"]
        botResponse += "\n"
        if (questionsCorrectCount != 1):
            botResponse += "You got {0} questions correct out of ".format(questionsCorrectCount)
        else:
            botResponse += "You got 1 question correct out of "
        botResponse += "{0}, with a score of ".format(questionsAnsweredCount)
        botResponse += "{0:.2f}%. ".format((questionsCorrectCount * 100.0)/(questionsAnsweredCount * 1.0))
        botResponse += "Thank you for taking the quiz! "
        event["res"]["session"].pop("quizBot", None)
        event["res"]["session"].pop("queryLambda", None)
    
    event["res"]["message"] = botResponse
    print(json.dumps(event))
    return event
    
def getDocumentContents(documentID, event):
    print("getting document contents for: ")
    print(documentID)
    
    client = boto3.client("lambda")
    tempResponse = client.invoke(
        FunctionName = event["req"]["_info"]["es"]["service"]["qid"],
        Payload = json.dumps({"qid":documentID}),
        InvocationType = "RequestResponse"
        )
    tempResponsePopulated = tempResponse["Payload"].read()
    return json.loads(tempResponsePopulated)

def quizShouldContinue(event):
    current=event["res"]["session"].get("quizBot", {}).get("current", "")
    first=event["res"]["result"].get("args")[0]
    wantsToQuit=userWantsToQuit(event["req"]["question"])
    return (current or first) and not wantsToQuit

def userWantsToQuit(userAnswer):
    tempStr = standardizeString(userAnswer)
    tempBool = ((tempStr == "QUIT") or (tempStr == "EXIT"))
    return tempBool

def getNextQuestionQid(quizBotDict):
    nextQid = quizBotDict["next"]
    
    if (isinstance(quizBotDict["next"], list) and len(quizBotDict["next"]) > 0):
        return nextQid[0]
    else:
        return nextQid
    
def standardizeString(inputString):
    translationTable = str.maketrans({key: None for key in string.punctuation})
    st = inputString.translate(translationTable)
    return st.upper().strip()

def assembleAnswerOptions(documentContents):
    allAnswers = []
    for answer in documentContents.get("incorrectAnswers", []):
        answerAndCorrectness = [answer, False]
        allAnswers.append(answerAndCorrectness)
    
    for answer in documentContents.get("correctAnswers", []):
        answerAndCorrectness = [answer, True]
        allAnswers.append(answerAndCorrectness)
       
    random.shuffle(allAnswers)
    
    letterCode = ord('A')
    for answerAndCorrectness in allAnswers:
        answerAndCorrectness[0] = chr(letterCode) + ". " + answerAndCorrectness[0]
        letterCode += 1
    
    return allAnswers

